internal/pkg/template/repository/manager/manager.go

```diff
@@ -0,0 +1,251 @@
+	// This list is individually modified in the Service according to the set project features.
+	defaultRepositories []model.TemplateRepository
+
+	repositories     map[string]*CachedRepository
+	repositoriesInit *singleflight.Group // each template is load only once
+	repositoriesLock *sync.RWMutex       // provides atomic access to the repositories field
+}
+
+type dependencies interface {
+	Tracer() trace.Tracer
+	Logger() log.Logger
+	ServerWaitGroup() *sync.WaitGroup
+}
+
+func New(ctx context.Context, defaultRepositories []model.TemplateRepository, d dependencies) (*Manager, error) {
+	m := &Manager{
+		ctx:                 ctx,
+		deps:                d,
+		logger:              d.Logger(),
+		defaultRepositories: defaultRepositories,
+		repositories:        make(map[string]*CachedRepository),
+		repositoriesInit:    &singleflight.Group{},
+		repositoriesLock:    &sync.RWMutex{},
+	}
+
+	// Free all repositories on server shutdown
+	serverWg := d.ServerWaitGroup()
+	serverWg.Add(1)
+	go func() {
+		defer serverWg.Done()
+		<-ctx.Done()
+		m.Free()
+	}()
+
+	// Init default repositories in parallel.
+	// It preloads all default repositories and templates in them.
+	errors := utils.NewMultiError()
+	initWg := &sync.WaitGroup{}
+	for _, repo := range defaultRepositories {
+		repo := repo
+		initWg.Add(1)
+		go func() {
+			defer initWg.Done()
+			if _, err := m.repository(ctx, repo); err != nil {
+				errors.Append(err)
+			}
+		}()
+	}
+	initWg.Wait()
+	return m, errors.ErrorOrNil()
+}
+
+// DefaultRepositories returns list of default repositories configured for the API.
+func (m *Manager) DefaultRepositories() []model.TemplateRepository {
+	return m.defaultRepositories
+}
+
+// ManagedRepositories return list of git repositories which are updated when the Pull method is called.
+func (m *Manager) ManagedRepositories() (out []string) {
+	m.repositoriesLock.RLock()
+	defer m.repositoriesLock.RUnlock()
+	for _, r := range m.repositories {
+		out = append(out, r.String())
+	}
+	sort.Strings(out)
+	return out
+}
+
+func (m *Manager) Repository(ctx context.Context, ref model.TemplateRepository) (*CachedRepository, UnlockFn, error) {
+	cachedRepo, err := m.repository(ctx, ref)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Prevented cleaning of the repository if it is outdated but still in use by some API request.
+	// UnlockFn must be called when the repository is no longer used.
+	unlockFn := cachedRepo.markInUse()
+	return cachedRepo, unlockFn, nil
+}
+
+func (m *Manager) Update(ctx context.Context) <-chan error {
+	errorCh := make(chan error, 1)
+	errors := utils.NewMultiError()
+
+	// Lock repositories field (during reading it in following for cycle)
+	m.repositoriesLock.Lock()
+
+	// Pull repositories in parallel
+	wait := &sync.WaitGroup{}
+	for _, repo := range m.repositories {
+		oldValue := repo
+		wait.Add(1)
+		go func() {
+			defer wait.Done()
+
+			newValue, updated, err := oldValue.update(ctx)
```

Comment:
Pri pravidelnom update sa zavola `ClonedRepository.update`, co vrati novu verziu repository (ak sa nasla zmeny, iny commit hash).

Stare requesty dobehnu so starou verziou, nove uz bezia s novou verziou (podobne ako to bolo pre FS).